File: src/tools.py
"""Local toolbox for the Doubao-TUI AI Agent.

Each public function is a callable tool that the LLM can invoke via OpenAI
function-calling.  Every tool returns a descriptive ``str`` (success or error)
so the model can understand the outcome and adjust.

The module also exports ``TOOL_SCHEMAS`` — the JSON Schema definitions required
by the OpenAI ``tools`` parameter.
"""

from __future__ import annotations

import os
import subprocess
from typing import Any, Dict, List

# ---------------------------------------------------------------------------
# Excluded directory names for view_structure
# ---------------------------------------------------------------------------
_EXCLUDE_DIRS: set[str] = {
    ".git",
    "__pycache__",
    ".venv",
    "venv",
    "node_modules",
    ".mypy_cache",
    ".pytest_cache",
    ".tox",
    ".eggs",
    "dist",
    "build",
    ".idea",
    ".vscode",
}


# ===================================================================
# Tool implementations
# ===================================================================

def view_structure() -> str:
    """Return a tree-like listing of the current working directory.

    Excludes common VCS / build / virtual-environment directories.

    Returns:
        An indented tree string showing the project layout.
    """
    cwd: str = os.getcwd()
    lines: List[str] = [f"{cwd}/"]
    try:
        for root, dirs, files in os.walk(cwd):
            # Prune excluded directories in-place
            dirs[:] = [d for d in dirs if d not in _EXCLUDE_DIRS]

            level = root.replace(cwd, "").count(os.sep)
            indent = "    " * level
            basename = os.path.basename(root) or root
            if level > 0:
                lines.append(f"{indent}├── {basename}/")

            sub_indent = "    " * (level + 1)
            for fname in sorted(files):
                lines.append(f"{sub_indent}├── {fname}")
    except OSError as exc:
        return f"Error walking directory: {exc}"

    return "\n".join(lines)


def read_file(path: str) -> str:
    """Read and return the full contents of a file.

    Args:
        path: Relative or absolute path to the target file.

    Returns:
        The file contents as a string, or an error description.
    """
    try:
        with open(path, "r", encoding="utf-8", errors="replace") as fh:
            return fh.read()
    except FileNotFoundError:
        return f"Error: file not found — {path}"
    except PermissionError:
        return f"Error: permission denied — {path}"
    except IsADirectoryError:
        return f"Error: {path} is a directory, not a file"
    except OSError as exc:
        return f"Error reading file: {exc}"


def write_file(path: str, content: str) -> str:
    """Overwrite or create a file with the given content.

    Parent directories are created automatically when missing.

    Args:
        path: Relative or absolute path to the target file.
        content: The full text to write.

    Returns:
        A success message or error description.
    """
    try:
        dirname = os.path.dirname(path)
        if dirname:
            os.makedirs(dirname, exist_ok=True)
        with open(path, "w", encoding="utf-8") as fh:
            fh.write(content)
        return f"Successfully wrote {len(content)} bytes to {path}"
    except OSError as exc:
        return f"Error writing file: {exc}"


def patch_file(path: str, old_text: str, new_text: str) -> str:
    """Perform an exact string replacement inside a file (first occurrence only).

    Args:
        path: Path to the target file.
        old_text: The exact substring to replace.
        new_text: The replacement string.

    Returns:
        A success message or error description.
    """
    try:
        with open(path, "r", encoding="utf-8", errors="replace") as fh:
            original = fh.read()
    except FileNotFoundError:
        return f"Error: file not found — {path}"
    except OSError as exc:
        return f"Error reading file for patch: {exc}"

    if old_text not in original:
        return (
            f"Error: old_text not found in {path}. "
            f"Ensure the snippet matches exactly (whitespace, indentation)."
        )

    patched = original.replace(old_text, new_text, 1)
    try:
        with open(path, "w", encoding="utf-8") as fh:
            fh.write(patched)
        return f"Successfully patched {path} (replaced 1 occurrence)"
    except OSError as exc:
        return f"Error writing patched file: {exc}"


def execute_command(command: str) -> str:
    """Execute a shell command and return stdout + stderr.

    The command runs with a 30-second timeout.  This tool is **high-risk** and
    must pass the security blacklist check + user authorisation before execution.

    Args:
        command: The shell command to run.

    Returns:
        Combined stdout and stderr (truncated at 4 000 chars), or an error message.
    """
    try:
        proc = subprocess.run(
            command,
            shell=True,
            capture_output=True,
            text=True,
            timeout=30,
        )
    except subprocess.TimeoutExpired:
        return f"Error: command timed out after 30 s — {command!r}"
    except FileNotFoundError:
        return f"Error: command not found — {command!r}"
    except OSError as exc:
        return f"Error executing command: {exc}"

    output = proc.stdout.strip() or "(no stdout)"
    if proc.stderr.strip():
        output += "\n[stderr]\n" + proc.stderr.strip()

    max_chars = 4_000
    if len(output) > max_chars:
        output = output[:max_chars] + f"\n\n... (truncated, {len(output)} chars total)"

    return f"[exit code {proc.returncode}]\n{output}"


# ===================================================================
# OpenAI Tool JSON Schemas
# ===================================================================

TOOL_SCHEMAS: List[Dict[str, Any]] = [
    {
        "type": "function",
        "function": {
            "name": "view_structure",
            "description": (
                "Show the directory tree of the current project. "
                "Use this to understand the file layout before reading or editing."
            ),
            "parameters": {
                "type": "object",
                "properties": {},
                "required": [],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "read_file",
            "description": (
                "Read the entire contents of a specified file. "
                "Returns the text or an error message."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "path": {
                        "type": "string",
                        "description": "Relative or absolute path to the file to read.",
                    },
                },
                "required": ["path"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "write_file",
            "description": (
                "Create a new file or overwrite an existing one with the given content. "
                "Parent directories are created automatically. "
                "**Requires user approval** before execution."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "path": {
                        "type": "string",
                        "description": "Relative or absolute path to the file.",
                    },
                    "content": {
                        "type": "string",
                        "description": "The full text content to write into the file.",
                    },
                },
                "required": ["path", "content"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "patch_file",
            "description": (
                "Replace an exact substring inside a file (first match only). "
                "Use this for surgical edits.  The old_text must match exactly "
                "(including whitespace and indentation).  **Requires user approval**."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "path": {
                        "type": "string",
                        "description": "Path to the file to patch.",
                    },
                    "old_text": {
                        "type": "string",
                        "description": "Exact substring to replace.",
                    },
                    "new_text": {
                        "type": "string",
                        "description": "Replacement string.",
                    },
                },
                "required": ["path", "old_text", "new_text"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "execute_command",
            "description": (
                "Run a shell command in the terminal and return its output. "
                "Commands are screened against a security blacklist. "
                "**Requires user approval** before execution.  "
                "Timeout: 30 seconds."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "command": {
                        "type": "string",
                        "description": "The shell command to execute.",
                    },
                },
                "required": ["command"],
            },
        },
    },
]

# Mapping used by the agent for dispatch
TOOL_DISPATCH: Dict[str, Any] = {
    "view_structure": view_structure,
    "read_file": read_file,
    "write_file": write_file,
    "patch_file": patch_file,
    "execute_command": execute_command,
}

File: src/tools.js
// ---------------------------------------------------------------------------
// ArkTerm — Tools: 5 local tool implementations + OpenAI function schemas
// ---------------------------------------------------------------------------
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const chalk = require('chalk');
const inquirer = require('inquirer');
const { isSafeCommand } = require('./security');

const WORKSPACE = process.cwd();

// ── Tool implementations ──────────────────────────────────────────────────

/**
 * View the directory structure of the workspace (depth ≤ 3).
 */
function viewStructure(rootDir) {
  const root = rootDir || WORKSPACE;
  const lines = [];
  lines.push(`📁 ${root}`);

  function walk(dir, prefix, depth) {
    if (depth > 3) return;
    let entries;
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    // Sort: dirs first, then alphabetical
    entries.sort((a, b) => {
      if (a.isDirectory() !== b.isDirectory())
        return a.isDirectory() ? -1 : 1;
      return a.name.localeCompare(b.name);
    });

    for (let i = 0; i < entries.length; i++) {
      const entry = entries[i];
      if (entry.name.startsWith('.') || entry.name === 'node_modules') continue;
      const isLast = i === entries.length - 1;
      const connector = isLast ? '└── ' : '├── ';
      lines.push(prefix + connector + entry.name + (entry.isDirectory() ? '/' : ''));
      if (entry.isDirectory()) {
        const nextPrefix = prefix + (isLast ? '    ' : '│   ');
        walk(path.join(dir, entry.name), nextPrefix, depth + 1);
      }
    }
  }

  walk(root, '', 0);
  return lines.join('\n');
}

/**
 * Read a file (UTF-8) and return its content head/tail.
 */
function readFile(filePath) {
  const abs = path.resolve(WORKSPACE, filePath);
  if (!fs.existsSync(abs)) {
    return `[Error] 文件不存在: ${filePath}`;
  }
  const stat = fs.statSync(abs);
  if (!stat.isFile()) {
    return `[Error] 不是文件: ${filePath}`;
  }
  const content = fs.readFileSync(abs, 'utf-8');
  const lines = content.split('\n');
  const totalLines = lines.length;
  const totalChars = content.length;

  // Preview: first 50 + last 50 lines if too large
  if (totalLines > 200) {
    const head = lines.slice(0, 50).join('\n');
    const tail = lines.slice(-50).join('\n');
    return (
      `📄 ${filePath} (${totalLines} lines, ${totalChars} chars)\n` +
      `─── HEAD (1-50) ───\n${head}\n` +
      `─── ... (${totalLines - 100} lines omitted) ───\n` +
      `─── TAIL (${totalLines - 49}-${totalLines}) ───\n${tail}`
    );
  }

  return `📄 ${filePath} (${totalLines} lines, ${totalChars} chars)\n${content}`;
}

/**
 * Write (or overwrite) a file. Returns a confirmation string.
 * Supports both full content and patching instructions.
 */
function writeFile(filePath, content) {
  const abs = path.resolve(WORKSPACE, filePath);
  const dir = path.dirname(abs);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
  fs.writeFileSync(abs, content, 'utf-8');
  const stat = fs.statSync(abs);
  return `✅ ${filePath} written (${stat.size} bytes).`;
}

/**
 * Apply a text replacement patch to an existing file.
 * Simulates the original Python `patch_file` logic.
 */
function patchFile(filePath, search, replace) {
  const abs = path.resolve(WORKSPACE, filePath);
  if (!fs.existsSync(abs)) {
    return `[Error] 文件不存在: ${filePath}`;
  }
  const original = fs.readFileSync(abs, 'utf-8');
  if (!original.includes(search)) {
    // try fuzzy: strip leading whitespace tolerance
    const normalized = original
      .split('\n')
      .map((l) => l.trimEnd())
      .join('\n');
    const searchNorm = search
      .split('\n')
      .map((l) => l.trimEnd())
      .join('\n');
    if (normalized.includes(searchNorm)) {
      // Reconstruct with original whitespace — just do direct replacement
      return `[Error] 精确匹配失败，存在空白符差异。请提供精确的原文片段。`;
    }

    // Try to locate similar lines
    const searchFirstLine = search.split('\n')[0].trim();
    const lines = original.split('\n');
    const similar = lines
      .map((l, i) => ({ line: l, index: i }))
      .filter(({ line }) => line.includes(searchFirstLine.slice(0, 30)));
    if (similar.length > 0) {
      const hints = similar
        .slice(0, 3)
        .map(({ line, index }) => `  L${index + 1}: ${line.slice(0, 80)}`)
        .join('\n');
      return (
        `[Error] 未找到精确匹配。附近行:\n${hints}\n` +
        `请提供精确的原文片段（注意缩进格式）`
      );
    }

    return `[Error] 未找到匹配内容: "${search.slice(0, 60)}..."`;
  }

  const updated = original.replace(search, replace);
  if (updated === original) {
    return `[Error] 替换后无变化，请检查 search/replace 内容。`;
  }

  fs.writeFileSync(abs, updated, 'utf-8');
  const replacedLen = search.length;
  return `✅ ${filePath} patched (${replacedLen} chars replaced).`;
}

/**
 * Execute a shell command (with security gate).
 */
async function executeCommand(command) {
  // ── Windows command translation ─────────────────────────────────────────
  if (process.platform === 'win32') {
    // ls → dir: strip -flags that don't map to Windows /flags
    command = command.replace(/^ls(\s+.*)?$/i, (_, trail) => {
      if (!trail) return 'dir';
      const cleaned = trail.replace(/(?:^|\s)--?\w+/g, '').replace(/\s+/g, ' ').trim();
      return cleaned ? `dir ${cleaned}` : 'dir';
    });
    // cat → type: same treatment
    command = command.replace(/^cat(\s+.*)?$/i, (_, trail) => {
      if (!trail) return 'type';
      const cleaned = trail.replace(/(?:^|\s)--?\w+/g, '').replace(/\s+/g, ' ').trim();
      return cleaned ? `type ${cleaned}` : 'type';
    });
  }

  // ── Security gate ──────────────────────────────────────────────────────
  if (!isSafeCommand(command)) {
    return `[Blocked] 命令被安全模块拦截 (匹配黑名单)。`;
  }

  console.log(chalk.yellow(`\n  [Security Warning] Executing: ${command}`));
  const { confirmed } = await inquirer.prompt([{
    type: 'confirm',
    name: 'confirmed',
    message: 'Execute this command?',
    default: false,
  }]);
  if (!confirmed) {
    return `[Denied] 用户未批准该命令。`;
  }

  // ── Execution ──────────────────────────────────────────────────────────
  try {
    const stdout = execSync(command, {
      cwd: WORKSPACE,
      timeout: 30_000,
      maxBuffer: 10 * 1024 * 1024,
      encoding: 'utf-8',
      windowsHide: true,
    });
    const output = stdout || '(no output)';
    return `$ ${command}\n${output.slice(0, 3000)}${
      output.length > 3000 ? '\n… (output truncated)' : ''
    }`;
  } catch (err) {
    const stderr = err.stderr || err.message || '(unknown error)';
    const stdout = err.stdout || '';
    return `$ ${command}\n${chalk.red('Error:')} ${stderr.slice(0, 1000)}${
      stdout ? `\n${stdout.slice(0, 1000)}` : ''
    }`;
  }
}

// ── OpenAI Function‑Calling Schemas ───────────────────────────────────────

const TOOL_SCHEMAS = [
  {
    type: 'function',
    function: {
      name: 'view_structure',
      description:
        '查看工作区的目录结构（递归，最大深度 3 层）。排除 . 开头和 node_modules 目录。',
      parameters: {
        type: 'object',
        properties: {
          path: {
            type: 'string',
            description: '可选的子路径，默认为工作区根目录',
          },
        },
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'read_file',
      description: '读取工作区内的一个 UTF‑8 文件并返回内容预览（大文件只返回首尾）。',
      parameters: {
        type: 'object',
        properties: {
          path: {
            type: 'string',
            description: '文件路径（相对于工作区根目录）',
          },
        },
        required: ['path'],
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'write_file',
      description: '创建或覆盖写入一个文件（自动创建父目录）。',
      parameters: {
        type: 'object',
        properties: {
          path: {
            type: 'string',
            description: '文件路径（相对于工作区根目录）',
          },
          content: {
            type: 'string',
            description: '文件完整内容',
          },
        },
        required: ['path', 'content'],
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'patch_file',
      description:
        '对已有文件执行精确的搜索‑替换操作。search 必须与原文精确匹配（包括缩进）。',
      parameters: {
        type: 'object',
        properties: {
          path: {
            type: 'string',
            description: '文件路径（相对于工作区根目录）',
          },
          search: {
            type: 'string',
            description: '要被替换的原文片段（必须精确匹配，包括缩进）',
          },
          replace: {
            type: 'string',
            description: '替换后的新内容',
          },
        },
        required: ['path', 'search', 'replace'],
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'execute_command',
      description:
        '在操作系统 Shell 中执行命令。会自动检查安全黑名单并征求用户批准。超时 30 秒。',
      parameters: {
        type: 'object',
        properties: {
          command: {
            type: 'string',
            description: '要执行的 Shell 命令',
          },
        },
        required: ['command'],
      },
    },
  },
];

// ── Dispatch table ────────────────────────────────────────────────────────

const TOOL_DISPATCH = {
  view_structure: async (args) => viewStructure(args.path || WORKSPACE),
  read_file: async (args) => readFile(args.path),
  write_file: async (args) => writeFile(args.path, args.content),
  patch_file: async (args) => patchFile(args.path, args.search, args.replace),
  execute_command: async (args) => executeCommand(args.command),
};

module.exports = {
  TOOL_SCHEMAS,
  TOOL_DISPATCH,
};
